File: src/main/java/me/diamondforge/kyromera/levelcardlib/UserData.java
package me.diamondforge.kyromera.levelcardlib;

/**
 * Class that holds all user data needed to generate a level card.
 * Uses the builder pattern for flexible and readable instantiation.
 */
public class UserData {
    private final String username;
    private final int rank;
    private final int level;
    private final int minXP;
    private final int maxXP;
    private final int currentXP;
    private final byte[] avatarBytes;
    private final String avatarUrl;
    private final ImageMode imageMode;
    private final OnlineStatus onlineStatus;
    private final boolean showStatusIndicator;

    private UserData(Builder builder) {
        this.username = builder.username;
        this.rank = builder.rank;
        this.level = builder.level;
        this.minXP = builder.minXP;
        this.maxXP = builder.maxXP;
        this.currentXP = builder.currentXP;
        this.avatarBytes = builder.avatarBytes;
        this.avatarUrl = builder.avatarUrl;
        this.imageMode = builder.imageMode;
        this.onlineStatus = builder.onlineStatus;
        this.showStatusIndicator = builder.showStatusIndicator;
    }

    /**
     * Gets the username.
     *
     * @return The username
     */
    public String getUsername() {
        return username;
    }

    /**
     * Gets the user's rank.
     *
     * @return The rank
     */
    public int getRank() {
        return rank;
    }

    /**
     * Gets the user's level.
     *
     * @return The level
     */
    public int getLevel() {
        return level;
    }

    /**
     * Gets the minimum XP for the current level.
     *
     * @return The minimum XP
     */
    public int getMinXP() {
        return minXP;
    }

    /**
     * Gets the maximum XP for the current level.
     *
     * @return The maximum XP
     */
    public int getMaxXP() {
        return maxXP;
    }

    /**
     * Gets the user's current XP.
     *
     * @return The current XP
     */
    public int getCurrentXP() {
        return currentXP;
    }

    /**
     * Gets the avatar image as a byte array.
     * May be null if using a URL.
     *
     * @return The avatar image bytes or null
     */
    public byte[] getAvatarBytes() {
        return avatarBytes;
    }

    /**
     * Gets the avatar URL.
     * May be null if using local image bytes.
     *
     * @return The avatar URL or null
     */
    public String getAvatarUrl() {
        return avatarUrl;
    }

    /**
     * Gets the image mode (LOCAL or DOWNLOAD).
     *
     * @return The image mode
     */
    public ImageMode getImageMode() {
        return imageMode;
    }

    /**
     * Gets the user's online status.
     *
     * @return The online status
     */
    public OnlineStatus getOnlineStatus() {
        return onlineStatus;
    }

    /**
     * Checks if the status indicator should be shown.
     *
     * @return true if the status indicator should be shown, false otherwise
     */
    public boolean isShowStatusIndicator() {
        return showStatusIndicator;
    }

    /**
     * Builder class for UserData.
     */
    public static class Builder {
        private String username;
        private int rank = 1;
        private int level = 1;
        private int minXP = 0;
        private int maxXP = 100;
        private int currentXP = 0;
        private byte[] avatarBytes;
        private String avatarUrl;
        private ImageMode imageMode = ImageMode.LOCAL;
        private OnlineStatus onlineStatus = OnlineStatus.ONLINE;
        private boolean showStatusIndicator = true;

        /**
         * Creates a new Builder with the required username.
         *
         * @param username The username (required)
         */
        public Builder(String username) {
            if (username == null || username.isEmpty()) {
                throw new IllegalArgumentException("Username cannot be null or empty");
            }
            this.username = username;
        }

        /**
         * Sets the user's rank.
         *
         * @param rank The rank
         * @return This builder for chaining
         */
        public Builder rank(int rank) {
            if (rank < 1) {
                throw new IllegalArgumentException("Rank must be positive");
            }
            this.rank = rank;
            return this;
        }

        /**
         * Sets the user's level.
         *
         * @param level The level
         * @return This builder for chaining
         */
        public Builder level(int level) {
            if (level < 1) {
                throw new IllegalArgumentException("Level must be positive");
            }
            this.level = level;
            return this;
        }

        /**
         * Sets the XP values.
         *
         * @param minXP The minimum XP for the current level
         * @param maxXP The maximum XP for the current level
         * @param currentXP The user's current XP
         * @return This builder for chaining
         */
        public Builder xp(int minXP, int maxXP, int currentXP) {
            if (minXP < 0 || maxXP <= minXP || currentXP < minXP || currentXP > maxXP) {
                throw new IllegalArgumentException(
                    "Invalid XP values: minXP must be >= 0, maxXP > minXP, and minXP <= currentXP <= maxXP");
            }
            this.minXP = minXP;
            this.maxXP = maxXP;
            this.currentXP = currentXP;
            return this;
        }

        /**
         * Sets the avatar image from a byte array.
         * This will set the image mode to LOCAL.
         *
         * @param avatarBytes The avatar image as a byte array
         * @return This builder for chaining
         */
        public Builder avatarBytes(byte[] avatarBytes) {
            this.avatarBytes = avatarBytes;
            this.imageMode = ImageMode.LOCAL;
            return this;
        }

        /**
         * Sets the avatar image URL.
         * This will set the image mode to DOWNLOAD.
         *
         * @param avatarUrl The URL of the avatar image
         * @return This builder for chaining
         */
        public Builder avatarUrl(String avatarUrl) {
            this.avatarUrl = avatarUrl;
            this.imageMode = ImageMode.DOWNLOAD;
            return this;
        }

        /**
         * Sets the online status.
         *
         * @param onlineStatus The online status
         * @return This builder for chaining
         */
        public Builder onlineStatus(OnlineStatus onlineStatus) {
            this.onlineStatus = onlineStatus;
            return this;
        }

        /**
         * Sets whether to show the status indicator.
         *
         * @param showStatusIndicator true to show the indicator, false to hide it
         * @return This builder for chaining
         */
        public Builder showStatusIndicator(boolean showStatusIndicator) {
            this.showStatusIndicator = showStatusIndicator;
            return this;
        }

        /**
         * Builds the UserData instance.
         *
         * @return A new UserData instance
         */
        public UserData build() {
            // Validate that we have either avatar bytes or URL
            if (imageMode == ImageMode.LOCAL && avatarBytes == null) {
                throw new IllegalStateException("Avatar bytes must be provided when using LOCAL image mode");
            }
            if (imageMode == ImageMode.DOWNLOAD && (avatarUrl == null || avatarUrl.isEmpty())) {
                throw new IllegalStateException("Avatar URL must be provided when using DOWNLOAD image mode");
            }
            
            return new UserData(this);
        }
    }
}

File: src/main/kotlin/me/diamondforge/kyromera/levelcardlib/UserData.kt
package me.diamondforge.kyromera.levelcardlib

/**
 * Class that holds all user data needed to generate a level card.
 * Uses the builder pattern for flexible and readable instantiation.
 */
data class UserData private constructor(
    val username: String,
    val discriminator: String,
    val level: Int,
    val rank: Int,
    val currentXP: Int,
    val nextLevelXP: Int,
    val messagesCount: Int,
    val voiceTime: String,
    val streak: Int,
    val avatarBytes: ByteArray?,
    val avatarUrl: String?,
    val imageMode: ImageMode
) {
    // Override equals and hashCode because ByteArray uses reference equality by default
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as UserData

        if (username != other.username) return false
        if (discriminator != other.discriminator) return false
        if (level != other.level) return false
        if (rank != other.rank) return false
        if (currentXP != other.currentXP) return false
        if (nextLevelXP != other.nextLevelXP) return false
        if (messagesCount != other.messagesCount) return false
        if (voiceTime != other.voiceTime) return false
        if (streak != other.streak) return false
        if (avatarBytes != null) {
            if (other.avatarBytes == null) return false
            if (!avatarBytes.contentEquals(other.avatarBytes)) return false
        } else if (other.avatarBytes != null) return false
        if (avatarUrl != other.avatarUrl) return false
        if (imageMode != other.imageMode) return false

        return true
    }

    override fun hashCode(): Int {
        var result = username.hashCode()
        result = 31 * result + discriminator.hashCode()
        result = 31 * result + level
        result = 31 * result + rank
        result = 31 * result + currentXP
        result = 31 * result + nextLevelXP
        result = 31 * result + messagesCount
        result = 31 * result + voiceTime.hashCode()
        result = 31 * result + streak
        result = 31 * result + (avatarBytes?.contentHashCode() ?: 0)
        result = 31 * result + (avatarUrl?.hashCode() ?: 0)
        result = 31 * result + imageMode.hashCode()
        return result
    }

    /**
     * Builder class for UserData.
     */
    class Builder(private val username: String) {
        init {
            require(username.isNotEmpty()) { "Username cannot be null or empty" }
        }

        private var discriminator: String = ""
        private var rank: Int = 1
        private var level: Int = 1
        private var currentXP: Int = 0
        private var nextLevelXP: Int = 100
        private var messagesCount: Int = 0
        private var voiceTime: String = ""
        private var streak: Int = 0
        private var avatarBytes: ByteArray? = null
        private var avatarUrl: String? = null
        private var imageMode: ImageMode = ImageMode.LOCAL

        /**
         * Sets the user's discriminator (the 4-digit number after the username).
         *
         * @param discriminator The discriminator
         * @return This builder for chaining
         */
        fun discriminator(discriminator: String): Builder {
            this.discriminator = discriminator
            return this
        }

        /**
         * Sets the user's rank.
         *
         * @param rank The rank
         * @return This builder for chaining
         */
        fun rank(rank: Int): Builder {
            require(rank > 0) { "Rank must be positive" }
            this.rank = rank
            return this
        }

        /**
         * Sets the user's level.
         *
         * @param level The level
         * @return This builder for chaining
         */
        fun level(level: Int): Builder {
            require(level >= 0) { "Level must be positive" }
            this.level = level
            return this
        }

        /**
         * Sets the XP values.
         *
         * @param currentXP The user's current XP
         * @param nextLevelXP The XP required for the next level
         * @return This builder for chaining
         */
        fun xp(currentXP: Int, nextLevelXP: Int): Builder {
            require(currentXP >= 0 && nextLevelXP > 0 && currentXP <= nextLevelXP) {
                "Invalid XP values: currentXP must be >= 0, nextLevelXP > 0, and currentXP <= nextLevelXP"
            }
            this.currentXP = currentXP
            this.nextLevelXP = nextLevelXP
            return this
        }

        /**
         * Sets the user's message count.
         *
         * @param messagesCount The number of messages
         * @return This builder for chaining
         */
        fun messagesCount(messagesCount: Int): Builder {
            require(messagesCount >= 0) { "Messages count must be non-negative" }
            this.messagesCount = messagesCount
            return this
        }

        /**
         * Sets the user's voice time.
         *
         * @param voiceTime The voice time as a formatted string (e.g., "127h")
         * @return This builder for chaining
         */
        fun voiceTime(voiceTime: String): Builder {
            this.voiceTime = voiceTime
            return this
        }

        /**
         * Sets the user's streak.
         *
         * @param streak The streak count
         * @return This builder for chaining
         */
        fun streak(streak: Int): Builder {
            require(streak >= 0) { "Streak must be non-negative" }
            this.streak = streak
            return this
        }

        /**
         * Sets the avatar image from a byte array.
         * This will set the image mode to LOCAL.
         *
         * @param avatarBytes The avatar image as a byte array
         * @return This builder for chaining
         */
        fun avatarBytes(avatarBytes: ByteArray): Builder {
            this.avatarBytes = avatarBytes
            this.imageMode = ImageMode.LOCAL
            return this
        }

        /**
         * Sets the avatar image URL.
         * This will set the image mode to DOWNLOAD.
         *
         * @param avatarUrl The URL of the avatar image
         * @return This builder for chaining
         */
        fun avatarUrl(avatarUrl: String): Builder {
            this.avatarUrl = avatarUrl
            this.imageMode = ImageMode.DOWNLOAD
            return this
        }


        /**
         * Builds the UserData instance.
         *
         * @return A new UserData instance
         */
        fun build(): UserData {
            // Validate that we have either avatar bytes or URL
            when (imageMode) {
                ImageMode.LOCAL -> require(avatarBytes != null) {
                    "Avatar bytes must be provided when using LOCAL image mode"
                }
                ImageMode.DOWNLOAD -> require(!avatarUrl.isNullOrEmpty()) {
                    "Avatar URL must be provided when using DOWNLOAD image mode"
                }
            }

            return UserData(
                username,
                discriminator,
                level,
                rank,
                currentXP,
                nextLevelXP,
                messagesCount,
                voiceTime,
                streak,
                avatarBytes,
                avatarUrl,
                imageMode
            )
        }
    }
}
